text/0000-deployment-language.md

```diff
@@ -0,0 +1,561 @@
+- Tremor Issue: [tremor-rs/tremor-runtime#0000](https://github.com/tremor-rs/tremor-runtime/issues/0000)
+- RFC PR: [tremor-rs/tremor-rfcs#0000](https://github.com/tremor-rs/tremor-rfcs/pull/0000)
+
+# Summary
+[summary]: #summary
+
+Add a deployment language to tremor with which whole event flows, consisting of sinks, sources and pipelines can be defined, instantiated, run and removed again.
+
+# Motivation
+[motivation]: #motivation
+
+The main reason is that current deployments are split in several parts:
+
+1. sink/source (onramp/offramp) definition
+2. pipeline definition in trickle
+3. linking definition which connects 1. to 2.
+4. mapping definition, which instatiates and finally deploys the connected parts from 3.
+
+We have to split those across at least 2 files: 1 yaml file for artefacts, linking and mapping, 1 trickle file for the pipeline.
+
+It would be much nicer for users to define their deployment in one unit, one file.
+Also, given our experiences with tremor-script and trickle, a DSL that is well suited for the task at hand can provide benefits such as:
+
+* helpful feedback on errors
+* modularization, outsourcing common definitions to reusable modules (e.g. HTTP proxy setup with onramps, offramps, pipeline)
+* more expressive
+* makes it possible to introduce a REPL for issuing language statements against a tremor installation (be it a single server or a cluster)
+
+The main advantage is that we can integrate the whole pipeline and script definitions seamlessly and reuse language concepts and idioms, so users will
+get a look and feel of the deployment language, trickle and tremor-script to be 1 language with slight extensions based on context (e.g. pipeline definition).
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A full tremor deployment always consists of one or more `onramps`, connected to one or more `pipelines` connected to one or more `offramps`.
+These are deployed as a single unit that describes how these artefacts are connected.
+
+The lifecycle of an artefact during deployment is (in very verbose terms) the following:
+
+* publish its definition (with all configuration) via unique (per artefact type) artefact id
+  - publishes the artefact definition to the repository under the artefact id
+  - makes it resolvable via a Tremor URL
+* create an instance from the artefact referenced by its type and id (or via Tremor URL) with a unique (per artefact) instance id
+  - instantiates the artefact and publishes it to the registry under the artefact-id and instance-id
+  - makes it resolvable via a Tremor URL
+* link the instance to other compatible instances (e.g. pipeline -> offramp or onramp -> pipeline), thus creating a connected and actively running event flow
+
+There are also the following steps:
+
+* Disconnect an artefact instance from all its connections (e.g. disconnect an offramp from its pipelines)
+* Destroy an artefact instance and thus stop its operation
```

Comment:
We might want to mention that removing instances trickles down (up?) to connected instances for automatic reclemation of resources. And if / how we do that in the future